### DISCUSSION TITLE
Disable tree shaking

### DISCUSSION THREAD

**Discussion by DreamTexX** [swc-project/swc#2880] [Q&A]
Hey, is it possible to prevent the transformer from removing "unnessecary" imports from typescript when transforming it to javascript?

**Comment by kdy1** [comment]
Are you using swc from js, there's no api to do so.

**Reply by DreamTexX** [reply]
yes i use it from js, can you add an api for this?

**Reply by kwonoj** [reply]
Mind share usecases? Curious to hear if it is something suppose to be exposed as public interface.

**Reply by DreamTexX** [reply]
Well, my current usecase would be something like this:
I've written a little parser for a vue like component file:
```vue
<script lang="ts">
  import Comonent from "./some-other-component";
</script>

<template>
  <Component></Component>
</template>
```
Now, i want to use swc to parse the content of the script tag, which is typescript, into js. But because "Component" is not directly used by the script it removes it, but it should keep it for transformations that i do later.

**Reply by DreamTexX** [reply]
So... will the api for this be added?

**Reply by kdy1** [reply]
Considering you want js api, you want to strip types without lowering them? Or you want to just parse it?

**Reply by DreamTexX** [reply]
It should just parse the ts content into js, without removing imports. I don't bother if it checks if the ts code is correct or not

**Reply by kdy1** [reply]
Then I think you can use `swc.parse()`?

**Reply by DreamTexX** [reply]
I've tried but how do i get js output from that?

**Reply by kdy1** [reply]
You can use `swc.print` but I don't think it will be performant enough.

**Reply by DreamTexX** [reply]
Well i currently parse it to get all imports, than i transform it and the I add the imports back into the new ast... so it propably would be much more efficient than the current... the simplest solution would be to add an option do disable tree shaking or something like this

**Reply by kwonoj** [reply]
Would you mind try suggested approach first and see if it works? For now, it's little bit hard to determine if this feature is something frequent that core api surface need to support as public for long run.

**Reply by DreamTexX** [reply]
Well i've tried, eighter i do something wrong or it doesn work, but when i load the script with swc.parse and the print it with swc.print it just keep the typings (even when targeting ecmascript). So i switched to parsing it first and getting the import nodes that must be kept, than i transform the code with swc.transform, then i parse the transformed code again and check if the desired imports are present. If not i add them and swc.print the stuff again. This is way to complicated, and i don't really see the point of not adding it to the api as it does not bother anyone who does not need it. If it is already a present feature in the crates i don't get why not publishing it, but rather forcing me to use an inefficient workaround.

**Reply by kwonoj** [reply]
I don't think I'm in position to make a decision solely, but in my opnion to support / expand api surface, I think there's firm need with frequent usecases. Having public api surface to opt in obviously won't hurt any user, but in terms of core team's maintanence we have to keep up across semver, reason I mentioned `bit hard to determine if this feature is something frequent that core api surface needs`. 

To double clarify, I'm not trying to say either we shouldn't or should support this. Also I don't represent whole core team's opinoin. It is my personal opinion it may need some actual usecases to be part of public api surface.
